Feature/enable weighted voting

p4u opened the pull request.

ed255 committed:
Detect too many pre-registers in rolling census

Keep track of used weight per user in the pre-register phase.  For now
we only accept weight 1 for each RegisterKeyTx; but in theory it should
be already possible to pre-register several keys (with weight 1 each) if
the pre-register census is weighted, for example with an ERC20 census.

Resolve https://github.com/vocdoni/vocdoni-node/issues/333

p4u committed:
enable weighted voting

Final tweaks to enable OFF_CHAIN_MERKLE_TREE weighted voting.

The census service will now return the voter value (weight)
on genProof, so the voter can use it when casting the vote.

The vochain checks the provided value if the census is Arbo
based (before a weight of 1 was hardcoded).

Extend integration tests in order to support weight.

Signed-off-by: p4u <pau@dabax.net>

arnaucube commented on client/api.go:
```diff
@@ -837,7 +844,7 @@ func (c *Client) TestSendAnonVotes(
 		return 0, err
 	}
-	log.Infof("CircuitIndex: %v, CircuitConfit: %+v", *circuitIndex, *circuitConfig)
+	log.Infof("CircuitIndex: %v, CircuitPath: %+v", *circuitIndex, *&circuitConfig.CircuitPath)
```
Why `*&circuitConfig.CircuitPath` and not `circuitConfig.CircuitPath`? as `circuit.Config.CircuitPath is a `string`.

```suggestion
	log.Infof("CircuitIndex: %v, CircuitPath: %+v", *circuitIndex, circuitConfig.CircuitPath)
```

p4u replied:
Thas vs-code autocomplete, I will change it, thanks

arnaucube commented on vochain/proof.go:
```diff
@@ -89,8 +89,12 @@ func VerifyProofOffChainTree(process *models.Process, proof *models.Proof,
+		if p.Value == nil {
+			return valid, bigOne, err
+		}
+		return valid, new(big.Int).SetBytes(p.Value), err
```
This is converting bytes to bigint interpreting bytes as big-endian, when those bytes come from inside the merkletree where those same bytes are treated as little-endian.
Maybe would make sense to update this line to return the bigint interepreting the bytes as little-endian, as are some bytes coming from inside the tree (which works in little-endian).

In fact, arbo already has a method that abstracts from 'endianness' to convert from bytes to bigint and viceversa, using that, this line could be:
```suggestion
		return valid, arbo.BytesToBigInt(p.Value), err
```

p4u replied:
So, there is something I don't understand.

In order to add the value, we do `err := tr.Add(data, r.CensusValue)`. The r.CensusValue is a slice of bytes. Are you saying that Arbo will switch the order of that slice internally on tr.Add() but when returning the value it will just return the bytes (without reversing them again)?

If that's the case I don't understand the logic behind it, thb.

arnaucube replied:
The thing is that if arbo is using a blake2b or sha256 hash internally, the endianness is not a problem. But, if the instance of arbo used is with Poseidon hash, arbo inputs are still `[]byte`, but later internally will need to be converted into `*big.Int` in order to compute the Poseidon hashes. So arbo interprets those byte arrays as little-endian in order to convert them to bigints (only in the Poseidon hash case): https://github.com/vocdoni/arbo/blob/master/tree.go#L278

Now, if we use the `Add` method without taking in account the endianness, we could input a byte-array that encodes a bigint in big-endian, and when internally is converted to bigint to compute the Poseidon hash it gives an error due the value being outside of the finite field.
As an example, the value `21888242871839275222246405745257275088548364400416034343698204186575808495616` (which is inside the finite field).
If we encode it as big-endian it is `0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593effff449` (in big-endian hex). If we give those bytes, internally when being interpreted as little-endian will give the value `33451715034712113185045486448192360878076553814222909500690766407339250443312`, which was not the original one, but that's not the only problem: when trying to compute the Poseidon hash of this value, it will give the error `inputs values not inside Finite Field`.

As an iteration on the previous example, if somebody sends a vote with `weight` equal to `11888242871839275222246405745257275088548364400416034343698204186575808499697` encoded as a big-endian byte array, when the Vochain tries to add it into a Poseidon Tree, it will not be able and it will return the mentioned error.

p4u committed:
client/api.go fix print variable

Co-authored-by: arnau <17317030+arnaucube@users.noreply.github.com>

mvdan commented on client/api.go:
```diff
@@ -222,7 +222,7 @@ func (c *Client) TestResults(pid []byte, totalVotes int) ([][]string, error) {
 		log.Infof("no results yet at block %d", block+2)
 	}
-	total := fmt.Sprintf("%d", totalVotes)
+	total := fmt.Sprintf("%d", uint64(totalVotes)*withWeight)
```
can this not overflow?

mvdan commented on urlapi/urlapi.go:
```diff
@@ -17,10 +17,11 @@ type URLAPI struct {
+	api         *bearerstdapi.BearerStandardAPI
+	scrutinizer *scrutinizer.Scrutinizer
+	vocapp      *vochain.BaseApplication
+	//lint:ignore U1000 unused
```
or just delete it? :)

p4u committed:
census api: use weight instead of censusValue

Let the clients deal with weight: a string representation of a big number.
Instead of the censusValue (raw byte slice) which requires a specific encoding.

Signed-off-by: p4u <pau@dabax.net>
